### PR TITLE
Assignment 4

### PR DESCRIPTION
**Pull request by ahmar-volopay**
Implemented Changes: 

- Added `Product` Button to the Dashboard.
- `Product` Page contains `Category` component to create filter/category for filtered Listing.
- Added required redux setups and files.

### CODE REVIEW

File: src/store/actions/productActions.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    dispatch(fetchProductRequest());
+    try {
+      const url = selectedCategory
+        ? `https://dummyjson.com/products/category/${selectedCategory}?limit=${limit}&skip=${skip}`
+        : `https://dummyjson.com/products?limit=${limit}&skip=${skip}`; // Default fetch if no category
+      const response = await axios.get(url);
+      
+      // Pass the replace flag along with the fetched data
+      dispatch(fetchProductSuccess({ products: response.data.products, replace }));
+      console.log(response.data);
```
Comment (gauravvolopay) [inline src/store/actions/productActions.js]: Remove console

File: src/components/core/Category.jsx [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+
+const Category = ({ name, onClick }) => {
```
Comment (gauravvolopay) [inline src/components/core/Category.jsx:4]: import PropTypes from "prop-types";
use **PropTypes** to ensure the correct datatype is passed for each prop

File: src/components/Products.jsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            setPage(0);               
+            setPrevCategory("");      
+            dispatch(fetchProduct(10, 0, "", true)); //Force Clear
+          }}
+        >
+          Clear
+        </button>
+      </div>
+
+      <table className="min-w-full table-auto border-collapse border border-gray-800">
```
Comment (gauravvolopay) [inline src/components/Products.jsx]: Use the same Table core component which is there in **core** folder.

File: src/components/Products.jsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  return (
+    <div className="p-4">
+      <h1 className="text-center font-bold text-2xl py-4">Product Table</h1>
+
+      <div className="p-2 flex items-center flex-wrap gap-2 font-semibold">
+        <div>Filter from Categories: </div>
+        {listCategory.length > 0
+          ? listCategory.slice(0, 5).map((category, index) => (
+            <Category
+              key={index}
```
Comment (gauravvolopay) [inline src/components/Products.jsx]: use category.slug or category.name  as key

File: src/components/Products.jsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+            />
+          ))
+          : "No categories available"}
+        <button
+          className="bg-gray-500 hover:bg-gray-700 text-white font-bold py-2 px-4 rounded"
+          onClick={() => {
+            setSelectedCategory("");  
+            setPage(0);               
+            setPrevCategory("");      
+            dispatch(fetchProduct(10, 0, "", true)); //Force Clear
```
Comment (gauravvolopay) [inline src/components/Products.jsx]: use the **limit** variable declared at the top 
image

File: src/components/Products.jsx [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  useEffect(() => {
+    const shouldReplace = selectedCategory !== prevCategory;
+    if (selectedCategory !== prevCategory) {
+      setPage(0);
+      setPrevCategory(selectedCategory);
+    }
+    dispatch(fetchProduct(limit, page * limit, selectedCategory, shouldReplace));
+  }, [dispatch, page, selectedCategory, prevCategory]);
+
+  const handleCategoryClick = (category) => {
```
Comment (gauravvolopay) [inline src/components/Products.jsx:59]: on click of a category, need to add query param as **?category=smartphones** in the url, currently not happening

**Review by gauravvolopay — CHANGES_REQUESTED** [review]

**Comment by gauravvolopay** [comment]
Repeatative main table api call is getting called

File: src/store/actions/productActions.js (hunk 1 above)
Comment (ahmar-volopay) [reply src/store/actions/productActions.js]: Resolved ✅

File: src/components/core/Category.jsx (hunk 2 above)
Comment (ahmar-volopay) [reply src/components/core/Category.jsx:4]: Resolved ✅

File: src/components/Products.jsx (hunk 3 above)
Comment (ahmar-volopay) [reply src/components/Products.jsx]: Resolved ✅

File: src/components/Products.jsx (hunk 4 above)
Comment (ahmar-volopay) [reply src/components/Products.jsx]: Resolved ✅

File: src/components/Products.jsx (hunk 5 above)
Comment (ahmar-volopay) [reply src/components/Products.jsx]: `limit` variable reused. Value updated to `15`.
Resolved ✅

File: src/components/Products.jsx (hunk 6 above)
Comment (ahmar-volopay) [reply src/components/Products.jsx:59]: Resolved ✅

**Review by gauravvolopay — APPROVED** [review]

**Pull request merged**